### PR TITLE
feat(tokenCreateTransaction): Implement `TokenCreateTransaction` E2E tests: TCK

### LINKED ISSUE #231: `TCK`: Implement tests for `TokenCreateTransaction`
Implement the method and tests specified in the documentation added by #139.

### PR DESCRIPTION
**Pull request by rwalworth**
**Description**:
This PR implements the `TokenCreateTransaction` tests documented in `test-specifications/token-service/tokenCreateTransaction.md`. Currently three tests are skipped because of a bug in the services code which allows for tokens to be created with a custom fee that collects to a deleted account. Another test is skipped because `deleteToken` is not yet implemented.

**Related issue(s)**:

Fixes #231 #244 

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: test/token-service/test_tokenCreateTransaction.js
Comment (ivaylogarnev-limechain) [inline test/token-service/test_tokenCreateTransaction.js:1]: As we continue to expand the test suite, we currently have around 150 tests for account transactions. With the PR for the token create transaction, the number of tests has increased to about 400. With the addition of other update/delete operations in the token transaction series, I anticipate the total will exceed 700 tests.

Some code is repeatable, as expected. A quick example is the Keys series of tests for TokenCreate transaction, which has about 2400 lines, where the tests are largely the same except for different key names being tested. As this scales over time, it becomes crucial to think about introducing **helper functions** or even **templates** to save time in future development.

Screenshot 2024-10-05 at 11 42 07

Additionally, some other tests across different transactions also show repeatable patterns, such as tests for the **Name** property or the **Auto Renew Period.** Introducing templates or other optimization techniques could streamline these tests as well.

File: test/token-service/test_tokenCreateTransaction.js [hunk 1]
```diff
@@ -0,0 +1,8187 @@
...
+      } catch (err) {
+        assert.equal(err.code, -32603, "Internal error");
+        return;
+      }
+
+      assert.fail("Should throw an error");
+    });
+  });
+
+  describe("Custom Fees", function () {
```
Comment (ivaylogarnev-limechain) [inline test/token-service/test_tokenCreateTransaction.js:3950]: We can: 
1. Extract the helper functions to different file in the **helpers** folder.
2. Merge common fee comparison logic with creating a generic helper functions to handle common comparison logic.
3. Write a single verifyTokenCreation function that handles the logic for comparing custom fees for different types, making it modular.
4. Call this general function for different fee types (Fixed, Fractional, Royalty) by providing the appropriate compare functions and types

The goal here is to enhance reusability and modularity, ensuring the code can be easily maintained and extended for future use.

File: test/token-service/test_tokenCreateTransaction.js [hunk 2]
```diff
@@ -0,0 +1,8187 @@
...
+
+// Needed to convert BigInts to JSON number format.
+BigInt.prototype.toJSON = function () {
+  return JSON.rawJSON(this.toString());
+};
+
+/**
+ * Tests for TokenCreateTransaction
+ */
+describe.only("TokenCreateTransaction", function () {
```
Comment (ivaylogarnev-limechain) [inline test/token-service/test_tokenCreateTransaction.js:27]: Delete .only, it was mistakenly forgotten.

File: utils/helpers/constants/token-type.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import { TokenSupplyType } from "@hashgraph/sdk";
```
Comment (ivaylogarnev-limechain) [inline utils/helpers/constants/token-type.js:1]: This should be TokenType, also please check why the "**Token Type**" test suite is passing.

File: test/token-service/test_tokenCreateTransaction.js [hunk 4]
```diff
@@ -0,0 +1,8187 @@
...
+      assert.fail("Should throw an error");
+    });
+
+    it("(#5) Creates a token with an auto renew period of 9,223,372,036,854,775,808 (`int64` max + 1) seconds", async function () {
+      try {
+        const response = await JSONRPCRequest("createToken", {
+          name: "testname",
+          symbol: "testsymbol",
+          treasuryAccountId: process.env.OPERATOR_ACCOUNT_ID,
+          autoRenewPeriod: 9223372036854775808n,
```
Comment (0xivanov) [inline test/token-service/test_tokenCreateTransaction.js:3080]: I am able to pass Duration of `9223372036854775808` seconds and it is passing in Java, how does it work in JS and c++ @ivaylogarnev-limechain 
 @rwalworth ?

Comment (ivaylogarnev-limechain) [reply test/token-service/test_tokenCreateTransaction.js:3080]: I haven't tested it in JS but in C++ its being handled by the server.

File: consensusInfoClient.js [hunk 5]
```diff
@@ -36,10 +36,19 @@ class ConsensusInfoClient {
     return this.executeAccountMethod(accountId, new AccountInfoQuery());
   }
 
+  async getTokenInfo(tokenId) {
+    return this.executeTokenMethod(tokenId, new TokenInfoQuery());
```
Comment (ivaylogarnev-limechain) [inline consensusInfoClient.js:40]: Import **TokenInfoQuery** here

File: test/token-service/test_tokenCreateTransaction.js [hunk 6]
```diff
@@ -0,0 +1,8187 @@
...
+        }
+      } catch (err) {
+        assert.equal(err.data.status, "INVALID_EXPIRATION_TIME");
+        return;
+      }
+
+      assert.fail("Should throw an error");
+    });
+
+    it("(#3) Creates a token with an expiration time of 9,223,372,036,854,775,807 (int64 max) seconds", async function () {
```
[the remark this answers is not in the document]
Comment (jsync-swirlds) [reply test/token-service/test_tokenCreateTransaction.js:2581]: Why would you need to create an Instant?  The field in the API is an integer, not an Instant.
This test is intended to verify that the SDK is correctly handling the edges of the integer value, and should be retained (in my opinion).

File: test/token-service/test_tokenCreateTransaction.js (hunk 4 above)
Comment (jsync-swirlds) [reply test/token-service/test_tokenCreateTransaction.js:3080]: hmmm, that value should be rejected by consensus as invalid; it parses to a negative number in the Java code.
The question is what is translating that Duration to an integer for the API, and is that SDK code applying a boundary limit.

Comment (0xivanov) [reply test/token-service/test_tokenCreateTransaction.js:3080]: Oh yes, it's actually overflowing.
The question now is now to pass a number that exceeds the int64 boundary as a int64 parameter :D

Comment (0xivanov) [reply test/token-service/test_tokenCreateTransaction.js:3080]: So we want it to overflow and the consensus would return an error because the value is negative?
Maybe some of the java util methods are doing some magic under the hood to cause weird behaviour, will check it.

File: test/token-service/test_tokenCreateTransaction.js [hunk 7]
```diff
@@ -0,0 +1,8187 @@
...
+        const response = await JSONRPCRequest("createToken", {
+          name: "",
+          symbol: "testsymbol",
+          treasuryAccountId: process.env.OPERATOR_ACCOUNT_ID,
+        });
+        if (response.status === "NOT_IMPLEMENTED") {
+          this.skip();
+        }
+      } catch (err) {
+        assert.equal(err.data.status, "MISSING_TOKEN_NAME");
```
[the remark this answers is not in the document]
Comment (ivaylogarnev-limechain) [reply test/token-service/test_tokenCreateTransaction.js:95]: It’s not a major concern, as we’ve been following this process up to this point.

File: test/token-service/test_tokenCreateTransaction.js [hunk 8]
```diff
@@ -0,0 +1,8187 @@
...
+        symbol: "testsymbol",
+        treasuryAccountId: process.env.OPERATOR_ACCOUNT_ID,
+        metadata: metadata,
+      });
+
+      if (response.status === "NOT_IMPLEMENTED") {
+        this.skip();
+      }
+
+      await verifyTokenCreationWithMetadata(response.tokenId, metadataValue);
```
[the remark this answers is not in the document]
Comment (ivaylogarnev-limechain) [reply test/token-service/test_tokenCreateTransaction.js:7873]: You're right, it just needs to be handled within the SDK server to convert it to bytes before passing it to the transaction.

File: test/token-service/test_tokenCreateTransaction.js (hunk 4 above)
Comment (jsync-swirlds) [reply test/token-service/test_tokenCreateTransaction.js:3080]: Yes.
Protobuf handles unsigned gracefully for Java (and we want the API to use unsigned for values that are never negative), but Java itself has no unsigned support, so the overflow case generally _should_ be tested and verified.

File: test/token-service/test_tokenCreateTransaction.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (0xivanov) [reply test/token-service/test_tokenCreateTransaction.js:2581]: The Java SDK currently has it's `expirationTime` as an `java.time.Instant`. 
This class has these constraints:
image

In theory it should support seconds of `int64 max value` but it actually does not...

File: test/token-service/test_tokenCreateTransaction.js [hunk 9]
```diff
@@ -0,0 +1,8187 @@
...
+        }
+      } catch (err) {
+        assert.equal(err.data.status, "INVALID_RENEWAL_PERIOD");
+        return;
+      }
+
+      assert.fail("Should throw an error");
+    });
+
+    it("(#5) Creates a token with an auto renew period of 9,223,372,036,854,775,808 (`int64` max + 1) seconds", async function () {
```
Comment (0xivanov) [inline test/token-service/test_tokenCreateTransaction.js:3074]: There is the following check in services:
image
where `NA = Long.MIN_VALUE;` and the autorenew period gets ignored in this test scenario.

File: test/token-service/test_tokenCreateTransaction.js [hunk 10]
```diff
@@ -0,0 +1,8187 @@
...
+      );
+    });
+
+    it("(#14) Creates a token with an expiration time of 8,000,002 seconds from the current time", async function () {
+      try {
+        const response = await JSONRPCRequest("createToken", {
+          name: "testname",
+          symbol: "testsymbol",
+          treasuryAccountId: process.env.OPERATOR_ACCOUNT_ID,
+          expirationTime: Date.now() / 1000 + 8000002,
```
Comment (0xivanov) [inline test/token-service/test_tokenCreateTransaction.js:2791]: this  number is of type `double` but it should not have floating point.

Comment (0xivanov) [reply test/token-service/test_tokenCreateTransaction.js:2791]: and after rounding the number the transaction goest through, but the assert is otherwise

File: test/token-service/test_tokenCreateTransaction.js [hunk 11]
```diff
@@ -0,0 +1,8187 @@
...
+        }
+      } catch (err) {
+        assert.equal(err.data.status, "INVALID_EXPIRATION_TIME");
+        return;
+      }
+
+      assert.fail("Should throw an error");
+    });
+
+    it("(#5) Creates a token with an expiration time of 9,223,372,036,854,775,808 (int64 max + 1) seconds", async function () {
```
Comment (0xivanov) [inline test/token-service/test_tokenCreateTransaction.js:2619]: same as https://github.com/hashgraph/hedera-sdk-tck/pull/252/files#r1794710905

Comment (0xivanov) [reply test/token-service/test_tokenCreateTransaction.js:2619]: Aren't the test cases with `int64 MAX +1` and `int64 MIN` the same? 
I mean I am handling the overflow of the numbers in my TCK server, because i expect an `Integer` and not `BigInteger`

``` java
        Optional<Long> parsedExpirationTime;
        try {
            parsedExpirationTime = Optional.ofNullable((Long) jrpcParams.get("expirationTime")); // handle long boundary
        } catch (Exception e) {
            parsedExpirationTime = Optional.of(((BigInteger) jrpcParams.get("expirationTime")).longValue()); // will overflow to negative number
        }

```

I am still a big confused if we really need `int64 MAX +1` cases.

File: test/token-service/test_tokenCreateTransaction.js (hunk 9 above)
Comment (ivaylogarnev-limechain) [reply test/token-service/test_tokenCreateTransaction.js:3074]: Yep its the same in JS, the test does not throw because period is getting ignored.

File: test/token-service/test_tokenCreateTransaction.js (hunk 10 above)
Comment (ivaylogarnev-limechain) [reply test/token-service/test_tokenCreateTransaction.js:2791]: We will modify how we handle the time values by converting them to ISO string format, as the current method for processing times is incorrect.

example for this specific test:
```
        const expirationDate = new Date(Date.now() + 8000002000);
        const expirationTimeIso = expirationDate.toISOString();
```
And we will send the **expirationTimeIso**

File: utils/helpers/constants/token-type.js (hunk 3 above)
Comment (rwalworth) [reply utils/helpers/constants/token-type.js:1]: I was trying to restrict use of the JS SDK as much as I could, I feel like since this is meant to test the JS SDK it should keep its dependence on it to a minimum. We could use another enum to track `"nft"` and `"ft"` token types, but the JSON sent across is a string so it must be one of these two options

File: test/token-service/test_tokenCreateTransaction.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (rwalworth) [reply test/token-service/test_tokenCreateTransaction.js:2581]: What is the reason this class enforces these constraints? Those seem like very arbitrary bounds

File: utils/helpers/constants/token-type.js (hunk 3 above)
Comment (ivaylogarnev-limechain) [reply utils/helpers/constants/token-type.js:1]: I agree that avoiding the use of the JS SDK and the TokenType class is the better approach. However, based on my personal experience, it might be a bit confusing for a new team member who has previously used TokenType to create a TokenCreateTransaction. Since it can be stringified using .toString(), it could still be used in this context. That said, I understand that simplifying things and steering clear of direct JS SDK usage will allow us to focus more on the tests rather than on the SDK itself.

In short, let's stick with the current approach.

File: test/token-service/test_tokenCreateTransaction.js (hunk 10 above)
Comment (ivaylogarnev-limechain) [reply test/token-service/test_tokenCreateTransaction.js:2791]: The problem occurs when we attempt to add or remove time using numbers alone. When we create a new Date() on the server (SDK) with these numeric values, it often results in an invalid date. This invalid date is then passed to the setter, causing an '**INVALID_EXPIRATION_TIME**' error. This error leads to the test passing, which is not the expected behavior. 

On the other hand, using ISO string format for the expiration time is a more reliable approach. Since ISO strings represent a standardized date format, they can't produce incorrect or invalid date values, ensuring that the expiration time is processed correctly while sending it as a string.

File: test/token-service/test_tokenCreateTransaction.js (hunk 11 above)
Comment (rwalworth) [reply test/token-service/test_tokenCreateTransaction.js:2619]: I believe the idea here was just to make sure the SDKs could parse both representations of the same underlying bit structure properly. I don't feel strongly either way about keeping or removing, @jsync-swirlds you have an opinion?

**Comment by rwalworth** [comment]
> Almost all the asserts for ecdsa keys are not passing because the consensus client is DER encoding the key differently than the Java SDK, I guess we could rework the logic to assert the keys raw? @rwalworth @ivaylogarnev-limechain WDYT?

I am okay with switching to just comparing raw keys, though we should do this for all key comparisons in all tests. I created an issue to track this https://github.com/hashgraph/hedera-sdk-tck/issues/257

File: test/token-service/test_tokenCreateTransaction.js (hunk 9 above)
Comment (jsync-swirlds) [reply test/token-service/test_tokenCreateTransaction.js:3077]: This seems like a bug in services code.  It should not be possible to set a value in the SDK that results in services ignoring the value.
If the services code needs to know if a value is or is not set in protobuf, we need to modify PBJ to support proto3 "explicit presence".  Currently an unset value just gets the field-type default (`0` in this case).

File: test/token-service/test_tokenCreateTransaction.js (hunk 11 above)
Comment (jsync-swirlds) [reply test/token-service/test_tokenCreateTransaction.js:2619]: We need these "out of bounds" edge case tests so that the TCK ensures that all SDKs handle these values (which are _valid_ in some languages) correctly and consistently.

Protocol Buffers do not translate these to `BigInteger` for Java, they translate as twos-compliment integers (so `INT64_MAX + 1` is, indeed, bitwise identical to `INT64_MIN`), but they do not translate the same in other SDK languages.
It seems, to me, that anywhere that different SDKs are expected to _correctly_ interpret the same protocol buffer serialized bytes in different ways are worth testing.

File: test/token-service/test_tokenCreateTransaction.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (ivaylogarnev-limechain) [reply test/token-service/test_tokenCreateTransaction.js:2581]: I believe this test should pass now that we've converted the BigInts to strings and pass **Long.fromString()** here.

File: utils/helpers/asn1-decoder.js
Comment (ivaylogarnev-limechain) [inline utils/helpers/asn1-decoder.js:1]: Please avoid checking this file, as the necessary updates should be handled [here](https://github.com/hashgraph/hedera-sdk-tck/pull/258).

File: test/token-service/test_tokenCreateTransaction.js (hunk 10 above)
Comment (rwalworth) [reply test/token-service/test_tokenCreateTransaction.js]: Understood, if the JS SDK is unable to parse a `BigInt` or `Long` into a `Date` then I'm okay with this move. Since we're changing the type of `expirationTime`, be sure to update the test specifications with the new type, as well as a description specifying it should be the ISO string format. I'll make an issue to change `expirationTime` parameters in other request types as well.

File: test/token-service/test_tokenCreateTransaction.js
Comment (rwalworth) [reply test/token-service/test_tokenCreateTransaction.js:1]: Created #260 to address this for all tests.

File: test/token-service/test_tokenCreateTransaction.js (hunk 1 above)
Comment (rwalworth) [reply test/token-service/test_tokenCreateTransaction.js:3994]: #260 can be used to investigate this.

File: test/token-service/test_tokenCreateTransaction.js (hunk 10 above)
Comment (rwalworth) [reply test/token-service/test_tokenCreateTransaction.js]: #259

File: test/token-service/test_tokenCreateTransaction.js (hunk 11 above)
Comment (0xivanov) [reply test/token-service/test_tokenCreateTransaction.js:2662]: But how is this testing the SDKs specifically when, for example, I have a method `setValueX(long value)` in my SDK and we are trying to "test" this method by passing value outside of the `long` boundary. 

I don't really see the benefit of doing this, as it only introduces overhead in the TCK server to "overflow the value" artificially.

Also, as far as I see, these tests are working in the C++ SDK because It has most (if not all) of it's `int64` values as `uint64`, see `amount`in https://github.com/hashgraph/hedera-sdk-cpp/blob/main/src/sdk/main/src/CustomFixedFee.cc , `maxSupply` in https://github.com/hashgraph/hedera-sdk-cpp/blob/main/src/sdk/main/src/TokenCreateTransaction.cc and many more. 
SDKs like Go, for example, have the above mentioned values as `int64` , so how can we make the behaviour of the TCK server consistent? 

I haven't ran the C++ TCK server but I'm pretty sure that when we pass `INT64_MAX + 1` it maps correctly because the parameters are `uint64` and not `int64` (or `long` in other languages) and then it's getting static casted like this `static_cast<int64_t>(mMaxSupply)` in order to overflow.

@rwalworth @jsync-swirlds

Comment (rwalworth) [reply test/token-service/test_tokenCreateTransaction.js:2662]: @jsync-swirlds what do you mean when you say "but they do not translate the same in other SDK languages."? Like an `int64` protobuf field wouldn't translate to a twos-complement signed integer type for another language? I think I agree with @0xivanov in that, at least for signed integer types, it seems a bit superfluous to test `INT64_MAX + 1` and `INT64_MIN` since they're the same underlying bit structure. Same with testing `UINT64_MAX` and `-1` for signed integer types. I can see the benefits of testing the `UINT64_MAX` values however if we have an unsigned protobuf value, solely because services doesn't support unsigned integers; I think that would be an overflow case that we'd want to test and make sure services handles properly.

File: test/token-service/test_tokenCreateTransaction.js [hunk 12]
```diff
@@ -0,0 +1,8256 @@
...
+          name: "testname",
+          symbol: "testsymbol",
+          treasuryAccountId: process.env.OPERATOR_ACCOUNT_ID,
+          customFees: [
+            {
+              feeCollectorAccountId: process.env.OPERATOR_ACCOUNT_ID,
+              feeCollectorsExempt: false,
+              fractionalFee: {
+                numerator: "1",
+                denominator: "9223372036854775808n",
```
Comment (rwalworth) [inline test/token-service/test_tokenCreateTransaction.js:4986]:
```suggestion
                denominator: "9223372036854775808",
```

**Comment by rwalworth** [comment]
Key comparison and large integer value changes are now working with the C++ SDK!

File: test/token-service/test_tokenCreateTransaction.js (hunk 11 above)
Comment (jsync-swirlds) [reply test/token-service/test_tokenCreateTransaction.js:2662]: Other languages interpret some protobuf types differently than Java (e.g. some languages may interpret `uint64` as a string, "Bignum", or similar, rather than twos-compliment binary, some _do_ support unsigned integers, etc...).  This is most pronounced for the unsigned types, not so much for signed types.
As a result of these differences it makes sense to test the edges of _signed integers_ for `uint64` and `uint32` values, including int_max+1 and int_min so that we have an opportunity to catch these differences.

int64_max+1 and int64_min are interesting in that they _are_ identical when stored unsigned or twos-compliment, but differ if stored as a string or Bignum, for instance, and that is why I suggest testing those two specifically.

Note, _in my opinion_ no SDK should be casting unsigned values to signed values that are _different_ from the protocol buffer definition; that would be a very strange behavior.

All edge cases are, in my opinion, suggestions; for any given test those edges might or might not be relevant.
If the protobuf type is `int32` or `int64`, then these edges don't make as much sense, of course.

Comment (rwalworth) [reply test/token-service/test_tokenCreateTransaction.js:2662]: @jsync-swirlds and I talked offline... for signed 64-bit types, we think testing something to the effect of `[0, -1, INT64_MAX - 1, INT64_MAX, INT64_MIN, INT64_MIN + 1]` as well as maybe one standard valid value should be sufficient. Does this sound good to you @0xivanov @ivaylogarnev-limechain?

**Review by gsstoykov — APPROVED** [review]

**Review by 0xivanov — APPROVED** [review]
